### PR TITLE
Aden-4399

### PR DESCRIPTION
**Pull request by bulaszenko**

### CODE REVIEW

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/adsbase/AdsBaseObject.java [hunk 1]
```diff
@@ -798,17 +794,19 @@ public void scrollToPosition(String selector) {
...
-    scrollToFooter();
-  }
 
-  public void clickOnAdImage(String slotName){
-    adSelector(slotName);
-    adSelector(slotName).click();
-    PageObjectLogging.log("clickOnAdImage", slotName + " is clicked", true);
+  // This scroll has been created because ad is not displayed if we scroll quickly to the Footer ADEN-4359
+  public void scrollToSlot(String slotName) {
+    switch(slotName) {
```
Comment (WikiaTech) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/adsbase/AdsBaseObject.java]: MINOR Replace this "switch" statement by "if" statements to increase readability. [rule](http://sonar-s1/coding_rules#rule_key=squid%3AS1301)
MAJOR Add a default case to this switch. [rule](http://sonar-s1/coding_rules#rule_key=squid%3ASwitchLastCaseIsDefaultCheck)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 1 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/ad/VideoFanTakeover.java [hunk 2]
```diff
@@ -1,19 +1,24 @@
...
 import org.openqa.selenium.WebElement;
+import org.openqa.selenium.interactions.Actions;
 
 public class VideoFanTakeover {
-  public static final double IMAGE_ASPECT_RATIO = 2.459;
-  public static final double VIDEO_ASPECT_RATIO = 1.769;
   private static final int VIDEO_LENGTH = 6000;
-  private static final By VIDEO_IFRAME_SELECTOR = By.cssSelector(".video-ima-container iframe");
-  private static By videoContainerSelector = By.cssSelector(".video-ima-container.hidden");
+  private static final String VIDEO_IFRAME_SELECTOR = "#%s .video-ima-container iframe";
```
Comment (nandy-andy) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/ad/VideoFanTakeover.java]: I don't know if we use this convention but it's up to you: there is a convention to use `_FORMAT` suffix when you have constants used as first parameter for `String.format()`, so in the examples here you'd have: `VIDEO_IFRAME_SELECTOR_FORMAT`, `MOBILE_VIDEO_SELECTOR_FORMAT` etc. It's up to you to use this convention or not.

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/ad/VideoFanTakeover.java [hunk 3]
```diff
@@ -1,19 +1,24 @@
...
+  private static final String MOBILE_VIDEO_SELECTOR = "#%s .video-ima-container video";
+  private static final String VIDEO_CONTAINER_SELECTOR = "#%s .video-ima-container.hidden";
+  private static final String UI_ELEMENT_SELECTOR = "#%s .overVideoLayer";
+  private static final String FANDOM_URL = "http://www.wikia.com/fandom";
+  private static final int PERCENT = 28;
```
Comment (nandy-andy) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/ad/VideoFanTakeover.java]: The name of this constant isn't very descriptive. Think of maybe a little bit longer but more descriptive name, please :)

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/ad/VideoFanTakeover.java [hunk 4]
```diff
@@ -61,29 +66,92 @@ private void runInAdFrame(Lambda f) {
...
-  public void waitForVideoStart() { waitForVideoPlayerVisible(); }
+  public void waitForVideoEnd(String slotName) {
+    wait.forElementPresent(By.cssSelector(String.format(VIDEO_CONTAINER_SELECTOR, slotName)), VIDEO_LENGTH);
+  }
 
-  public Double getCurrentVideoTime() {
+  public void clickOnAdImage(){
+    runInAdFrame(() -> {
+    Actions action = new Actions(driver);
+    action.moveToElement(wait.forElementClickable(playTriggerButtonSelector), -20, -20).click().build().perform();
```
Comment (nandy-andy) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/ad/VideoFanTakeover.java]: Where does these magic `-20`s come from? Do we have to use them?

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/ad/VideoFanTakeover.java [hunk 5]
```diff
@@ -61,29 +66,92 @@ private void runInAdFrame(Lambda f) {
...
+    });
+  }
+
+  public void waitForVideoStart(String slotName) { waitForVideoPlayerVisible(slotName); }
```
Comment (nandy-andy) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/ad/VideoFanTakeover.java]: Is it valid formatting?

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/ad/VideoFanTakeover.java [hunk 6]
```diff
@@ -61,29 +66,92 @@ private void runInAdFrame(Lambda f) {
...
 
-  public boolean isTimeProgressing(int quartileTime, int midTime) {
+  public Double getCurrentVideoTimeOnMobile(String slotName) {
+    String result;
+
+    result = driver.findElement(By.cssSelector(String.format(MOBILE_VIDEO_SELECTOR, slotName))).getAttribute("currentTime");
+    return Double.parseDouble(result);
+  }
+
+  public double getAdSlotHigh(String slotName) {
```
Comment (nandy-andy) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/ad/VideoFanTakeover.java]: `getAdSlotHigh()` -> `getAdSlotHeight()`

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/ad/VideoFanTakeover.java [hunk 7]
```diff
@@ -61,29 +66,92 @@ private void runInAdFrame(Lambda f) {
...
+    return driver.findElement(By.cssSelector(AdsContent.getSlotSelector(slotName))).getSize().getHeight();
+  }
+
+  public double getAdVideoHigh(String slotName) {
```
Comment (nandy-andy) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/ad/VideoFanTakeover.java]: `getAdVideoHigh()` -> `getAdVideoHeight()`

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/adsbase/AdsBaseObject.java [hunk 8]
```diff
@@ -798,17 +794,16 @@ public void scrollToPosition(String selector) {
...
-    scrollToFooter();
-    wait.forElementVisible(By.cssSelector(".editarea"));
-    scrollToFooter();
-  }
 
-  public void clickOnAdImage(String slotName){
-    adSelector(slotName);
-    adSelector(slotName).click();
-    PageObjectLogging.log("clickOnAdImage", slotName + " is clicked", true);
+  // This scroll has been created because ad is not displayed if we scroll quickly to the Footer ADEN-4359
```
Comment (nandy-andy) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/adsbase/AdsBaseObject.java]: I can't reproduce it :(

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/adsbase/AdsBaseObject.java [hunk 9]
```diff
@@ -798,17 +794,16 @@ public void scrollToPosition(String selector) {
...
+  public void scrollToSlot(String slotName) {
```
Comment (nandy-andy) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/adsbase/AdsBaseObject.java]: I don't think we need this method. Are there any reasons we can't wait for the slots being filled?

File: src/test/java/com/wikia/webdriver/testcases/adstests/TestAdsVUAP.java [hunk 10]
```diff
@@ -15,144 +15,137 @@
...
     videoFanTakeover.play();
-    videoFanTakeover.waitForVideoStart();
-    ads.verifySlotSize(slotName, ads.getViewPortWidth(), videoHeight);
-    videoFanTakeover.waitForVideoEnd();
-    ads.verifySlotSize(slotName, ads.getViewPortWidth(), imageHeight);
+
+    videoFanTakeover.waitForVideoStart(slotName);
+    double videoHeight = videoFanTakeover.getAdVideoHigh(slotName);
+
+    Assertion.assertTrue(videoFanTakeover.isVideoAdBiggerThanImageAdOasis(videoHeight, imageHeight ));
```
Comment (nandy-andy) [inline src/test/java/com/wikia/webdriver/testcases/adstests/TestAdsVUAP.java]: Unnecessary whitespace before `));`

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/ad/VideoFanTakeover.java (hunk 4 above)
Comment (bulaszenko) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/ad/VideoFanTakeover.java]: Yes I'm clicking here 20 pix left and 20 pix up from the button to make sure that i hit the ad image on mercury and oasis

File: src/test/java/com/wikia/webdriver/testcases/adstests/TestAdsVuapMercury.java [hunk 11]
```diff
@@ -1,36 +1,162 @@
...
 
     @NetworkTrafficDump(useMITM = true)
     @Test(
             dataProviderClass = AdsDataProvider.class,
-            dataProvider = "adsVUAPTopMercury",
+            dataProvider = "adsVuapMercury",
             groups = "AdsTopAdVideoClosesWhenFinishPlaysMercury"
     )
-    public void adsTopAdVideoClosesWhenFinishPlaysMercury(Page page, String slotName, String iframeId) {
+    public void adsTopAdVideoClosesWhenFinishPlaysMercury(Page page, String slotName, String iframeId, String videoUrl) {
```
Comment (nandy-andy) [inline src/test/java/com/wikia/webdriver/testcases/adstests/TestAdsVuapMercury.java]: I'd change the name here as well. Similar idea to the one above: `adsTestTopAdVideoBehavior()` or something like that.

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/ad/VideoFanTakeover.java (hunk 4 above)
Comment (nandy-andy) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/ad/VideoFanTakeover.java]: Let's document somehow or ignore and just click the button? :japanese_ogre: I see three options:
* just click the button and ignore the flexible margins around,
* leave a comment describing that the author chose 20px to make sure clicking outside the button works,
* move `-20` and assign it to a descriptive constant.

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/ad/VideoFanTakeover.java [hunk 12]
```diff
@@ -61,29 +66,91 @@ private void runInAdFrame(Lambda f) {
...
+  public void clickOnAdImage(){
+    runInAdFrame(() -> {
+    Actions action = new Actions(driver);
+    //Setting -20 value to make sure click will be executed 20px left and 20px up from video button to click on ad image
+    action.moveToElement(wait.forElementClickable(playTriggerButtonSelector), -20, -20).click().build().perform();
+    });
+  }
 
-  public void waitForVideoStart() { waitForVideoPlayerVisible(); }
+  public void clickOnVideoCloseButon() { wait.forElementVisible(closeVideoButton).click(); }
```
Comment (nandy-andy) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/ad/VideoFanTakeover.java]: Is this a valid formatting?

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/ad/VideoFanTakeover.java [hunk 13]
```diff
@@ -61,29 +66,91 @@ private void runInAdFrame(Lambda f) {
...
 
-  public boolean isTimeProgressing(int quartileTime, int midTime) {
+  public Double getCurrentVideoTimeOnMobile(String slotName) {
+    String result;
+
+    result = driver.findElement(By.cssSelector(String.format(MOBILE_VIDEO_SELECTOR_FORMAT, slotName))).getAttribute("currentTime");
+    return Double.parseDouble(result);
+  }
+
+  public double getAdSlotHeigh(String slotName) {
```
Comment (nandy-andy) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/ad/VideoFanTakeover.java]: `getAdSlotHeigh` -> `getAdSlotHeight`

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/ad/VideoFanTakeover.java [hunk 14]
```diff
@@ -61,29 +66,91 @@ private void runInAdFrame(Lambda f) {
...
+    return driver.findElement(By.cssSelector(AdsContent.getSlotSelector(slotName))).getSize().getHeight();
+  }
+
+  public double getAdVideoHeigh(String slotName) {
```
Comment (nandy-andy) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/ad/VideoFanTakeover.java]: `getAdVideoHeigh` -> `getAdVideoHeight`

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/adsbase/AdsBaseObject.java (hunk 9 above)
Comment (nandy-andy) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/pageobject/adsbase/AdsBaseObject.java]: We found today `triggerComments()` and used it here. But I'm still thinking we don't need `scrollToSlot()` especially it works only with two slots: bottom leaderboards (one slot per skin). Think about:
A) moving it into `TestAdsVUAP` class and re-use from them,
B) resigning from this method and using `scrollToElement()` with `triggerComments()` directly in the test's method.

File: src/test/java/com/wikia/webdriver/testcases/adstests/TestAdsVUAP.java [hunk 15]
```diff
@@ -18,141 +18,132 @@
 public class TestAdsVUAP extends TemplateNoFirstLoad {
```
Comment (nandy-andy) [inline src/test/java/com/wikia/webdriver/testcases/adstests/TestAdsVUAP.java]: Let's rename this class as well, so it's consistent with `TestAdsVuapMercury`.

Comment (bulaszenko) [inline src/test/java/com/wikia/webdriver/testcases/adstests/TestAdsVUAP.java]: class renamed to TestsAdsVuapOasis

File: src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/ad/VideoFanTakeover.java [hunk 16]
```diff
@@ -61,29 +66,93 @@ private void runInAdFrame(Lambda f) {
...
+  public boolean isVideoAdBiggerThanImageAdOasis(double videoHeight, double imageHeight) {
+    int percentResult = (int)Math.round(100-(100/(videoHeight/imageHeight)));
+    if (percentResult == PERCENTAGE_DIFFERENCE_BETWEEN_VIDEO_AND_IAMGE_AD) {
+      return true;
+    }
+    return false;
+  }
+
+// Different way of checking slot sizes on mercury because of the very small difference between two slots sizes
+  public boolean isVideoAdBiggerTahnImageAdMercury(double videoHeight, double imageHeight) {
```
Comment (nandy-andy) [inline src/test/java/com/wikia/webdriver/pageobjectsfactory/componentobject/ad/VideoFanTakeover.java]: `isVideoAdBiggerTahnImageAdMercury` -> `isVideoAdBiggerThanImageAdMercury`
